File: server/src/utils/cache.ts
import { createClient, RedisClientType } from 'redis';

class RedisCache {
  private client: RedisClientType | null = null;
  private isConnected: boolean = false;
  private clientCreated: boolean = false;

  constructor() {
    // Don't create Redis client in constructor - do it lazily
    console.log('Redis cache initialized (lazy loading)');
  }

  private async ensureClient(): Promise<void> {
    if (this.clientCreated) return;

    this.clientCreated = true;

    // Only create Redis client if Redis is available and not in test environment
    if (!process.env.REDIS_URL || process.env.NODE_ENV === 'test') {
      console.log('Redis not configured, running without cache');
      return;
    }

    try {
      this.client = createClient({
        url: process.env.REDIS_URL,
        socket: {
          connectTimeout: 5000, // Reduced timeout
        },
      });

      this.client.on('error', (err) => {
        console.warn('Redis connection error:', err.message);
        this.isConnected = false;
      });

      this.client.on('connect', () => {
        console.log('Connected to Redis');
        this.isConnected = true;
      });

      this.client.on('ready', () => {
        this.isConnected = true;
      });

      this.client.on('end', () => {
        this.isConnected = false;
      });
    } catch (error) {
      console.warn('Failed to create Redis client:', error);
      this.client = null;
    }
  }

  async connect(): Promise<void> {
    await this.ensureClient();
    if (!this.client || this.isConnected) return;

    try {
      await this.client.connect();
    } catch (error) {
      console.warn('Failed to connect to Redis, continuing without cache:', error);
    }
  }

  async disconnect(): Promise<void> {
    if (!this.client || !this.isConnected) return;

    try {
      await this.client.disconnect();
    } catch (error) {
      console.warn('Error disconnecting from Redis:', error);
    }
  }

  async get<T>(key: string): Promise<T | null> {
    if (!this.client || !this.isConnected) return null;

    try {
      const data = await this.client.get(key);
      return data ? JSON.parse(data) : null;
    } catch (error) {
      console.warn('Redis get error:', error);
      return null;
    }
  }

  async set(key: string, value: unknown, ttlSeconds?: number): Promise<void> {
    if (!this.client || !this.isConnected) return;

    try {
      const serializedValue = JSON.stringify(value);
      if (ttlSeconds) {
        await this.client.setEx(key, ttlSeconds, serializedValue);
      } else {
        await this.client.set(key, serializedValue);
      }
    } catch (error) {
      console.warn('Redis set error:', error);
    }
  }

  async delete(key: string): Promise<void> {
    if (!this.client || !this.isConnected) return;

    try {
      await this.client.del(key);
    } catch (error) {
      console.warn('Redis delete error:', error);
    }
  }

  async deletePattern(pattern: string): Promise<void> {
    if (!this.client || !this.isConnected) return;

    try {
      const keys = await this.client.keys(pattern);
      if (keys.length > 0) {
        await this.client.del(keys);
      }
    } catch (error) {
      console.warn('Redis delete pattern error:', error);
    }
  }

  async exists(key: string): Promise<boolean> {
    if (!this.client || !this.isConnected) return false;

    try {
      const result = await this.client.exists(key);
      return result === 1;
    } catch (error) {
      console.warn('Redis exists error:', error);
      return false;
    }
  }

  async clear(): Promise<void> {
    if (!this.client || !this.isConnected) return;

    try {
      await this.client.flushAll();
    } catch (error) {
      console.warn('Redis clear error:', error);
    }
  }

  // Cache wrapper for functions
  async cached<T>(
    key: string,
    fn: () => Promise<T>,
    ttlSeconds: number = 300
  ): Promise<T> {
    // Try to get from cache first
    const cached = await this.get<T>(key);
    if (cached !== null) {
      return cached;
    }

    // Execute function and cache result
    const result = await fn();
    await this.set(key, result, ttlSeconds);
    return result;
  }

  // Invalidate cache by pattern
  async invalidateUserCache(userId: string): Promise<void> {
    await this.deletePattern(`user:${userId}:*`);
  }

  async invalidateCurriculumCache(): Promise<void> {
    await this.deletePattern('curriculum:*');
  }

  async invalidateProgressCache(userId: string): Promise<void> {
    await this.deletePattern(`progress:${userId}:*`);
  }
}

// Export singleton instance
export const redisCache = new RedisCache();
export default redisCache;

File: ml-service/cache.py
"""
Redis caching utilities for ML service
"""
import redis
import json
import os
from typing import Any, Optional, Callable, TypeVar
from functools import wraps
import time

T = TypeVar('T')

class RedisCache:
    def __init__(self):
        self.redis_url = os.getenv('REDIS_URL', 'redis://localhost:6379')
        self.client = None
        self.is_connected = False

    def connect(self):
        """Connect to Redis with error handling"""
        try:
            self.client = redis.from_url(self.redis_url, decode_responses=True)
            self.client.ping()  # Test connection
            self.is_connected = True
            print("Connected to Redis")
        except redis.ConnectionError as e:
            print(f"Failed to connect to Redis: {e}")
            self.is_connected = False
        except Exception as e:
            print(f"Redis connection error: {e}")
            self.is_connected = False

    def disconnect(self):
        """Disconnect from Redis"""
        if self.client and self.is_connected:
            self.client.close()
            self.is_connected = False

    def get(self, key: str) -> Optional[Any]:
        """Get value from cache"""
        if not self.is_connected or not self.client:
            return None

        try:
            data = self.client.get(key)
            return json.loads(data) if data else None
        except Exception as e:
            print(f"Redis get error: {e}")
            return None

    def set(self, key: str, value: Any, ttl_seconds: int = 300) -> None:
        """Set value in cache with TTL"""
        if not self.is_connected or not self.client:
            return

        try:
            serialized_value = json.dumps(value)
            self.client.setex(key, ttl_seconds, serialized_value)
        except Exception as e:
            print(f"Redis set error: {e}")

    def delete(self, key: str) -> None:
        """Delete key from cache"""
        if not self.is_connected or not self.client:
            return

        try:
            self.client.delete(key)
        except Exception as e:
            print(f"Redis delete error: {e}")

    def delete_pattern(self, pattern: str) -> None:
        """Delete keys matching pattern"""
        if not self.is_connected or not self.client:
            return

        try:
            keys = self.client.keys(pattern)
            if keys:
                self.client.delete(*keys)
        except Exception as e:
            print(f"Redis delete pattern error: {e}")

    def exists(self, key: str) -> bool:
        """Check if key exists"""
        if not self.is_connected or not self.client:
            return False

        try:
            return bool(self.client.exists(key))
        except Exception as e:
            print(f"Redis exists error: {e}")
            return False

    def cached(self, ttl_seconds: int = 300, key_prefix: str = ""):
        """Decorator for caching function results"""
        def decorator(func: Callable[..., T]) -> Callable[..., T]:
            @wraps(func)
            def wrapper(*args, **kwargs) -> T:
                # Create cache key from function name and arguments
                key_parts = [key_prefix or func.__name__]
                key_parts.extend([str(arg) for arg in args if arg is not None])
                key_parts.extend([f"{k}:{v}" for k, v in kwargs.items() if v is not None])
                key = ":".join(key_parts)

                # Try to get from cache first
                cached_result = self.get(key)
                if cached_result is not None:
                    return cached_result

                # Execute function and cache result
                result = func(*args, **kwargs)
                self.set(key, result, ttl_seconds)
                return result

            return wrapper
        return decorator

    def invalidate_user_cache(self, user_id: str) -> None:
        """Invalidate all cache entries for a user"""
        self.delete_pattern(f"user:{user_id}:*")

    def invalidate_prediction_cache(self, user_id: str) -> None:
        """Invalidate prediction cache for a user"""
        self.delete_pattern(f"prediction:{user_id}:*")

    def invalidate_model_cache(self, model_name: str) -> None:
        """Invalidate cache for a specific model"""
        self.delete_pattern(f"model:{model_name}:*")


# Global cache instance
redis_cache = RedisCache()